lib/resque/plugins/stages.rb

```diff
@@ -15,41 +15,79 @@ class Error < StandardError; end
         if singleton_class.included_modules.map(&:name).include?("Resque::Plugins::Retry")
           try_again_callback :stages_report_try_again
           give_up_callback :stages_report_giving_up
-
-          add_record_inline_failure
         else
           add_record_failure
         end
       end
 
       # rubocop:disable Metrics/BlockLength
       class_methods do
+        def perform_job(*args)
+          job = perform_job_from_param(args)
+
+          if job.nil?
+            job            = Resque::Plugins::Stages::StagedJob.new(SecureRandom.uuid)
+            job.class_name = name
+            job.args       = args
+          end
+
+          job
+        end
+
+        # rubocop:disable Metrics/AbcSize
+        # rubocop:disable Metrics/CyclomaticComplexity
+        def perform_job_from_param(args)
+          return if args.blank? || !args.first.is_a?(Hash)
+
+          hash            = args.first.with_indifferent_access
+          job             = Resque::Plugins::Stages::StagedJob.new(hash[:staged_job_id]) if hash.key?(:staged_job_id)
+          job&.class_name = name
+          job.args        = (hash.key?(:resque_compressed) ? args : args[1..]) if !job.nil? && job.blank?
+
+          job
+        end
+
+        # rubocop:enable Metrics/CyclomaticComplexity
+        # rubocop:enable Metrics/AbcSize
+
         def before_perform_stages_successful(*args)
-          job = Resque::Plugins::Stages::StagedJob.perform_job(*args)
+          job = perform_job(*args)
 
           return if job.blank?
 
           job.status = :running
         end
 
         def after_perform_stages_successful(*args)
-          job = Resque::Plugins::Stages::StagedJob.perform_job(*args)
+          job = perform_job(*args)
 
           return if job.blank?
+          return if job.status == :failed && Resque.inline?
 
           job.status = :successful
         end
 
+        def around_perform_stages_inline_around(*args)
+          yield
+        rescue StandardError => e
+          raise e unless Resque.inline?
+
+          job = perform_job(*args)
+          return if job.blank?
+
+          job.status = :failed
```

Comment:
When performing in-line, we don't want an exception to be thrown, instead fail the job and continue.